### DISCUSSION TITLE
Extract query setFile() without storing the file on local storage

### DISCUSSION THREAD

**Discussion by gruenmap** [solariumphp/solarium#1066] [Q&A] [answered]
I am getting files from database in blob format and I want to send them to Solr Cell using extract query. However, the setFile() method in the Extract Query expects a local file path as input, but I want to avoid saving the binary file to the local storage before sending it to Solr instead I want to pass a binary file directly. Is it possible somehow to send a binary file directly with extract query or is there some better approach to this?

**Comment by thomascorthals** [comment]
I've opened an issue for this. It's not possible at the moment, but it's an easy fix.

**Reply by thomascorthals** [reply]
Working implementation in my fork: https://github.com/thomascorthals/solarium/tree/extract

I'll open a PR later. Would run into merge conflicts with another PR if I do it now.

**Reply by thomascorthals** [reply]
Caveat: my implementation has an issue with `extractOnly=true` that I'll need to fix first. Should be fine to use for indexing though.

**Reply by thomascorthals** [reply]
Got it figured out for `extractOnly=true`. It's even more merge conflicty now so you'll have to hang on for the PR.

**Comment by thomascorthals** [comment]
[Solarium 6.3.1](https://github.com/solariumphp/solarium/releases/tag/6.3.1) lets you extract from file pointer resources. This means you can now store data in memory instead of a file.

```php
$file = fopen('php://memory', 'w+');
fwrite($file, $blob);
$extract->setFile($file);

$client->extract($extract);

fclose($file);
```

Check the PHP manual on [php://memory and php://temp](https://www.php.net/manual/en/wrappers.php.php#wrappers.php.memory) for more details.

I think you can also use [PDO Large Objects](https://www.php.net/manual/en/pdo.lobs.php) to do it directly, but I haven't tested this.

**Reply by thomascorthals** [reply, marked as answer]
I tested it and this does work in PHP >= 8.1.0. It generates warnings with Solarium 6.3.1 though. I've opened PR #1083 to have this fixed in the next release. The PR also includes a full working example.

```php
$db = new PDO(...);

$select = $db->prepare("SELECT content FROM table WHERE id = ?");
$select->execute($id);
$select->bindColumn(1, $content, PDO::PARAM_LOB);
$select->fetch(PDO::FETCH_BOUND);

$query->setFile($content);

$client->extract($query);
```

**Reply by thomascorthals** [reply]
https://github.com/solariumphp/solarium/blob/master/examples/2.7.3.2-extract-query-pdo-lob.php

Should already work with a few PHP warnings, will work without those warnings in the next release.
